File: scripts/deprecated/precompact-save-chunk.py
#!/usr/bin/env python3
"""
Hook: PreCompact
Sauvegarde le transcript brut AVANT compaction pour éviter la perte d'information.
Le hook SessionEnd assemblera tous les chunks pour créer la mémoire complète.
"""
import json
import sys
import shutil
from pathlib import Path
from datetime import datetime

CHUNKS_DIR = Path.home() / ".claude" / "memory-chunks"


def save_transcript_chunk(transcript_path: Path, session_id: str, chunk_number: int):
    """Sauvegarde une copie du transcript avant compaction."""
    if not transcript_path.exists():
        return False

    try:
        # Créer le dossier de chunks
        CHUNKS_DIR.mkdir(parents=True, exist_ok=True)

        # Nom du chunk: session_id + numéro séquentiel
        chunk_filename = f"{session_id}_chunk_{chunk_number:03d}.jsonl"
        chunk_path = CHUNKS_DIR / chunk_filename

        # Copier le transcript
        shutil.copy2(transcript_path, chunk_path)

        # Sauvegarder un index pour cette session
        index_path = CHUNKS_DIR / f"{session_id}_index.json"
        index = []
        if index_path.exists():
            with open(index_path, 'r', encoding='utf-8') as f:
                index = json.load(f)

        index.append({
            "chunk_number": chunk_number,
            "filename": chunk_filename,
            "timestamp": datetime.now().isoformat(),
            "size": chunk_path.stat().st_size
        })

        with open(index_path, 'w', encoding='utf-8') as f:
            json.dump(index, f, indent=2)

        print(f"[precompact] Saved chunk {chunk_number} for session {session_id[:8]}", file=sys.stderr)
        return True

    except Exception as e:
        print(f"[precompact] Error saving chunk: {e}", file=sys.stderr)
        return False


def get_next_chunk_number(session_id: str) -> int:
    """Récupère le prochain numéro de chunk pour cette session."""
    index_path = CHUNKS_DIR / f"{session_id}_index.json"
    if not index_path.exists():
        return 1

    try:
        with open(index_path, 'r', encoding='utf-8') as f:
            index = json.load(f)
            if index:
                return max(c["chunk_number"] for c in index) + 1
    except Exception:
        pass

    return 1


def main():
    try:
        # Lire l'input du hook
        input_data = json.load(sys.stdin)
        transcript_path_str = input_data.get('transcript_path', '')
        session_id = input_data.get('session_id', 'unknown')

        if not transcript_path_str or session_id == 'unknown':
            sys.exit(0)

        transcript_path = Path(transcript_path_str)

        # Déterminer le numéro de chunk
        chunk_number = get_next_chunk_number(session_id)

        # Sauvegarder le chunk
        save_transcript_chunk(transcript_path, session_id, chunk_number)

        sys.exit(0)
    except Exception as e:
        print(f"[precompact] Error: {e}", file=sys.stderr)
        sys.exit(0)


if __name__ == '__main__':
    main()

File: scripts/deprecated/precompact-save-chunk.js
#!/usr/bin/env node
/**
 * Hook: PreCompact
 * Sauvegarde le transcript brut AVANT compaction pour éviter la perte d'information.
 * Le hook SessionEnd assemblera tous les chunks pour créer la mémoire complète.
 */

const fs = require('fs');
const path = require('path');
const os = require('os');

// Add lib directory to path for imports
const libDir = path.join(__dirname, 'lib');
const debugLogger = require(path.join(libDir, 'debug-logger.js'));

const { logHookStart, logHookEnd, logDebug } = debugLogger;

const CHUNKS_DIR = path.join(os.homedir(), '.claude', 'memory-chunks');

function saveTranscriptChunk(transcriptPath, sessionId, chunkNumber) {
    if (!fs.existsSync(transcriptPath)) {
        return false;
    }

    try {
        // Créer le dossier de chunks
        if (!fs.existsSync(CHUNKS_DIR)) {
            fs.mkdirSync(CHUNKS_DIR, { recursive: true });
        }

        // Nom du chunk: session_id + numéro séquentiel
        const chunkFilename = `${sessionId}_chunk_${String(chunkNumber).padStart(3, '0')}.jsonl`;
        const chunkPath = path.join(CHUNKS_DIR, chunkFilename);

        // Copier le transcript (fs.copyFileSync préserve les métadonnées)
        fs.copyFileSync(transcriptPath, chunkPath);

        // Sauvegarder un index pour cette session
        const indexPath = path.join(CHUNKS_DIR, `${sessionId}_index.json`);
        let index = [];

        if (fs.existsSync(indexPath)) {
            try {
                index = JSON.parse(fs.readFileSync(indexPath, 'utf-8'));
            } catch (e) {
                // Si l'index est corrompu, on recommence
                index = [];
            }
        }

        const stats = fs.statSync(chunkPath);
        index.push({
            chunk_number: chunkNumber,
            filename: chunkFilename,
            timestamp: new Date().toISOString(),
            size: stats.size
        });

        fs.writeFileSync(indexPath, JSON.stringify(index, null, 2), 'utf-8');

        logDebug('PreCompact', 'precompact-save-chunk.js', `Saved chunk ${chunkNumber} for session ${sessionId.substring(0, 8)}`, 'INFO');
        return true;

    } catch (e) {
        logDebug('PreCompact', 'precompact-save-chunk.js', `Error saving chunk: ${e.message}`, 'ERROR');
        return false;
    }
}

function getNextChunkNumber(sessionId) {
    const indexPath = path.join(CHUNKS_DIR, `${sessionId}_index.json`);

    if (!fs.existsSync(indexPath)) {
        return 1;
    }

    try {
        const data = fs.readFileSync(indexPath, 'utf-8');
        const index = JSON.parse(data);

        if (Array.isArray(index) && index.length > 0) {
            const maxChunk = Math.max(...index.map(c => c.chunk_number || 0));
            return maxChunk + 1;
        }
    } catch (e) {
        // Si erreur de lecture/parsing, recommencer à 1
    }

    return 1;
}

/* istanbul ignore next */
function main() {
    logHookStart('PreCompact', 'precompact-save-chunk.js');

    let inputData = '';
    process.stdin.setEncoding('utf8');

    process.stdin.on('readable', () => {
        let chunk;
        while ((chunk = process.stdin.read()) !== null) {
            inputData += chunk;
        }
    });

    process.stdin.on('end', () => {
        try {
            const data = JSON.parse(inputData);
            const transcriptPath = data.transcript_path || '';
            const sessionId = data.session_id || 'unknown';

            logDebug('PreCompact', 'precompact-save-chunk.js', `Transcript: ${transcriptPath}`, 'INFO');
            logDebug('PreCompact', 'precompact-save-chunk.js', `Session: ${sessionId}`, 'INFO');

            if (!transcriptPath || sessionId === 'unknown') {
                logDebug('PreCompact', 'precompact-save-chunk.js', 'Missing transcript or session_id', 'SKIP');
                logHookEnd('PreCompact', 'precompact-save-chunk.js', true);
                process.exit(0);
            }

            // Déterminer le numéro de chunk
            const chunkNumber = getNextChunkNumber(sessionId);
            logDebug('PreCompact', 'precompact-save-chunk.js', `Next chunk number: ${chunkNumber}`, 'INFO');

            // Sauvegarder le chunk
            const success = saveTranscriptChunk(transcriptPath, sessionId, chunkNumber);

            logDebug('PreCompact', 'precompact-save-chunk.js', success ? 'Chunk saved' : 'Chunk save failed', success ? 'INFO' : 'ERROR');
            logHookEnd('PreCompact', 'precompact-save-chunk.js', success);
            process.exit(0);

        } catch (e) {
            logDebug('PreCompact', 'precompact-save-chunk.js', `ERROR: ${e.message}`, 'ERROR');
            logHookEnd('PreCompact', 'precompact-save-chunk.js', false);
            process.exit(0);
        }
    });
}

// Export for testing
/* istanbul ignore else */
if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        saveTranscriptChunk,
        getNextChunkNumber,
        CHUNKS_DIR
    };
}

// Run main if executed directly
/* istanbul ignore if */
if (require.main === module) {
    main();
}
